packages/@lwc/ssr-compiler/src/compile-template/transformers/component/slotted-content.ts

```diff
@@ -62,6 +73,81 @@ const bAddLightContent = esTemplate`
     });
 `<EsCallExpression>;
 
+// Light DOM slots are a bit complex because of needing to handle slots _not_ at the top level
+// At the non-top level, it matters what the ancestors are. These are relevant to slots:
+// - Element/Text/Component/ExternalComponent (e.g. `<div>`, `<x-foo>`)
+// - If (`if:true`, `if:false`)
+// - IfBlock/ElseBlock/ElseifBlock (`lwc:if`, `lwc:elseif`, `lwc:else`)
+// Whereas anything else breaks the relationship between the slotted content and the containing
+// Component (e.g. another Component, an ExternalComponent, etc.), or is disallowed (e.g.
+// ForEach/ForOf).
+// The goal here is to traverse through the tree and identify all unique `slot` attribute names
+// and group those into AST trees on a per -`slot` name basis, only for ancestors that count
+// (as mentioned above).
+function getLightSlottedContent(rootNodes: IrChildNode[], cxt: TransformerContext) {
+    type SlottableAncestorIrType = IrElement | IrIf | IrIfBlock | IrElseBlock | IrElseifBlock;
+    type SlottableLeafIrType = IrElement | IrText | IrComponent | IrExternalComponent;
+    type SlottableIrType = SlottableAncestorIrType | SlottableLeafIrType;
+
+    const results: EsExpressionStatement[] = [];
+
+    // For the given slot name, get the EsExpressions we should use to render it
+    // The ancestorIndices is an array of integers referring to the chain of ancestors
+    // and their positions in the child arrays of their own parents
+    const addLightDomSlotContent = (slotName: EsExpression, ancestorIndices: number[]) => {
+        const clone = produce(rootNodes[ancestorIndices[0]], (draft) => {
+            // Create a clone of the AST tree with only the ancestors and no other siblings
+            let current = draft as SlottableIrType;
+            for (let i = 1; i < ancestorIndices.length; i++) {
+                const nextIndex = ancestorIndices[i];
+
+                // If i >= 1 then the current must necessarily be a SlottableAncestorIrType
+                const next = (current as SlottableAncestorIrType).children[nextIndex];
+                (current as SlottableAncestorIrType).children = [next];
+                current = next as SlottableIrType;
+            }
+            // The leaf must necessarily be a SlottableLeafIrType
+            const leaf = current as SlottableLeafIrType;
+            // Light DOM slots do not actually render the `slot` attribute.
+            if (leaf.type !== 'Text') {
+                leaf.attributes = leaf.attributes.filter((attr) => attr.name !== 'slot');
+            }
+        });
+        const slotContent = irToEs(clone, cxt);
+        results.push(b.expressionStatement(bAddLightContent(slotName, null, slotContent)));
+    };
```

Comment:
This logic is a bit hard to read, and the types are abysmal, but I'd put the PR up anyway and fix it later. Now you know how my twisted mind works. 😆